js/src/LogEvent.js

```diff
@@ -0,0 +1,126 @@
+(function(exports) {
+
+/**
+ * Elastic LogStash logging.
+ */
+class LogEvent
+{
+  constructor(url, { headers })
+  {
+    console.assert(url);
+    console.assert(headers);
+
+    this.url = url;
+    if (headers['set-cookie']) { // got set-cookie
+      let value = headers['set-cookie'];
+      let parts = value.split(';').filter((v) => !!v.trim());
+      let cookiePart = parts[0].split('=');
+      this.session = cookiePart[1];
+    }
+  }
+
+  /**
+   * Logs an error.
+   */
+  error(message)
+  {
+    this.event('error', { message });
+  }
+
+  /**
+   * Logs an event of the given type.
+   */
+  event(type, fields)
+  {
+    let data = Object.assign({
+      session: this.session,
+      type,
+      timestamp: Date.now()
+    }, fields);
+
+    let thisobj = this;
```

Comment:
```let self = this;``` is a bit shorter.

Reply:
I would avoid using self because this is a global property (https://developer.mozilla.org/en-US/docs/Web/API/Window/self)